nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessor.java

```diff
@@ -239,63 +251,155 @@ private WorkflowActionType getActionType(StateExecutionImpl execution) {
   }
 
   private boolean isNextActivationImmediately(StateExecutionImpl execution) {
-    return execution.getNextActivation() != null && !execution.getNextActivation().isAfterNow();
+    return execution.isStateProcessed() && execution.getNextActivation() != null && !execution.getNextActivation().isAfterNow();
   }
 
-  private NextAction processState(WorkflowInstance instance, AbstractWorkflowDefinition<?> definition,
-      StateExecutionImpl execution, WorkflowState currentState) {
-    WorkflowStateMethod method = definition.getMethod(instance.state);
-    if (method == null) {
-      execution.setNextState(currentState);
-      return stopInState(currentState, "Execution finished.");
+  private NextAction processWithListeners(ListenerContext listenerContext, final WorkflowInstance instance, final AbstractWorkflowDefinition<? extends WorkflowState> definition,
+                                          final StateExecutionImpl execution, final WorkflowState state) {
+    List<WorkflowExecutorListener> chain = new LinkedList<>(asList(this.executorListeners));
+    ProcessingExecutorListener processingListener = new ProcessingExecutorListener(instance, definition, execution, state);
+    chain.add(processingListener);
+    NextAction nextAction = new ExecutorListenerChain(chain).next(listenerContext);
```

Comment:
Could we create executor listener chain once when initializing the state processor?